src/test/coll.rs

```diff
@@ -1201,3 +1201,213 @@ fn assert_duplicate_key_error_with_utf8_replacement(error: &ErrorKind) {
+    let res = run_inserted_ids_test(&client, &docs, false);
+    let err = res.await.expect_err("insert_many should fail");
+    match *err.kind {
+        ErrorKind::BulkWrite(failure) => {
+            assert_eq!(
+                failure.inserted_ids.len(),
+                100000,
+                "100,000 documents should have been inserted"
+            );
+            // docs at index 1 up to and including 100,000 should have been inserted
+            for (i, doc) in docs.iter().enumerate().take(100001).skip(1) {
+                match failure.inserted_ids.get(&i) {
+                    Some(doc_id) => {
+                        let expected_id = doc.get("_id").unwrap();
+                        assert_eq!(
+                            doc_id, expected_id,
+                            "Doc at index {} did not have expected _id",
+                            i
+                        );
+                    }
+                    None => panic!("Document at index {} should have been inserted", i),
+                }
+            }
+        }
+        _ => panic!("Expected BulkWrite error, but got: {:?}", err),
+    }
+
+    // an ordered, 2-batch bulk write where the second-to-last document in the second batch
+    // generates a write error: everything before that document should be inserted
+    let mut docs = Vec::with_capacity(100002);
+    for i in 2..100003 {
+        docs.push(doc! { "_id": Bson::Int32(i) });
+    }
+    docs.push(doc! { "_id": 1 });
+    docs.push(doc! { "_id": 100003 });
+
+    let res = run_inserted_ids_test(&client, &docs, true);
+    let err = res.await.expect_err("insert_many should fail");
+    match *err.kind {
+        ErrorKind::BulkWrite(failure) => {
+            assert_eq!(
+                failure.inserted_ids.len(),
+                100001,
+                "100001 documents should have been inserted"
+            );
+            // docs at index 0 up to and including 100,000 should have been inserted;
+            // doc at index 100,001 generates a duplicate key error
+            for (i, doc) in docs.iter().enumerate().take(100001) {
+                match failure.inserted_ids.get(&i) {
+                    Some(doc_id) => {
+                        let expected_id = doc.get("_id").unwrap();
+                        assert_eq!(
+                            doc_id, expected_id,
+                            "Doc at index {} did not have expected _id",
+                            i
+                        );
+                    }
+                    None => panic!("Document at index {} should have been inserted", i),
+                }
+            }
+        }
+        _ => panic!("Expected BulkWrite error, but got: {:?}", err),
+    }
+
+    // an unordered, 2-batch bulk write where the second-to-last document in the second batch
+    // generates a write error: everything besides that document should be inserted
+    let res = run_inserted_ids_test(&client, &docs, false);
```

Comment:
Can we add one more case where we have a 2 batch write with a write error in both batches?